content/blog/2020-10-tokio-0-3.md

```diff
@@ -0,0 +1,138 @@
+
+Additionally, the `poll_read_buf` and `poll_write_buf` methods have been removed from both traits. In practice, these methods were rarely implemented. Vectored operations will be implemented by types that support them.
+
+## Feature flag simplification
+
+We have reduced the amount of feature flags needed to use the core features of Tokio. The `dns`, `tcp`, `udp`,  and `uds` feature flags are collapsed into a single `net` feature flag. We have also combined `rt-core` and `rt-util` into a single `rt` feature flag. This `rt` feature flag now encompasses everything to execute futures except for the multi-threaded runtime which now lives under a `rt-multi-thread` feature flag (previously was called `rt-threaded`).
+
+- `dns`, tcp`, `udp` and `uds` -> `net`
+- `rt-util` and `rt-core` -> `rt`
+- `rt-threaded` -> `rt-multi-threaded`
+
+## Runtime and Builder refactor
+
+The Tokio team has removed  `threaded_scheduler` and `basic_scheduler` from the public API. Instead, the runtime `Builder` type introduces two constructors: one constructor for a multi-threaded scheduler and one for a current thread scheduler. These constructors provide a misuse-resistant form of building a runtime that *does not* require fiddling with feature flags and builder methods. Tokio will no longer choose the runtime variant based on feature flags when using `Runtime::new` but instead will default to the multi-threaded runtime and is only available under the `rt` feature flag. In addition, we’ve renamed the `core_threads` builder method to the more accurate `worker_threads`. 
+
+We have also refactored the runtime module. Tokio 0.2 exposed two types to interact with the Tokio runtime: `Runtime` and `Handle`. `Runtime` has a `block_on` method that required `&mut self`, whereas `Handle` has a `block_on` method that required `&self`. With Tokio 0.3, we collapsed Runtime and `Handle` into a single `Runtime`. Tokio 0.3’s `Runtime`‘s `block_on` now accepts `&self`, which supports concurrent calls to `block_on`. This also means `Runtime` implements `Send` and `Sync` and can therefore be stored in an `Arc`.
+
+An example configuring a multi-threaded runtime with 6 worker threads:
+
+```rust
+use tokio::runtime::Builder;
+
+let rt = Builder::new_multi_thread()
+    .worker_threads(6)
+    .enable_all()
+    .build()
+    .unwrap();
+
+rt.block_on(async {
+    println!("Hello world!");
+});
+```
+
+## Methods changed to `&self`
```

Comment:
We should elaborate more on this. First, we should call out that the `&self` methods are for sockets, so maybe "Send and receive from a socket with `&self`. We should say it took an overhaul of the I/O driver. We should also explain why the trivial option doesn't work.

The problem comes w/ `poll_*` methods are supposed to only store the waker for the **last** call to `poll`. That doesn't support concurrent tasks calling.

> Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup.

https://doc.rust-lang.org/std/future/trait.Future.html

To make `&self` methods work, we had to restructure everything w/ intrusive wakers.